sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionManagerSample.ts

```diff
@@ -0,0 +1,155 @@
+    resolveExpirationPromise = res;
+  });
+
+  let timer = setTimeout(resolveExpirationPromise, timeoutMs);
+
+  return {
+    refresh() {
+      clearTimeout(timer);
+      timer = setTimeout(resolveExpirationPromise, timeoutMs);
+    },
+    expirationPromise
+  };
+}
+
+async function _noSessionsAvailable(err: Error) {
+  // TODO: this error should be programmatically identifable.
+  return err.message.search(/Received an incorrect sessionId 'undefined'/) !== -1;
+}
+
+async function _processNextSession(
+  serviceBusClient: ServiceBusClient,
+  handlers: SessionMessageHandlers,
+  abortSignal: AbortSignalLike
+): Promise<void> {
+  const sessionReceiver = serviceBusClient.createSessionReceiver(queueName, "peekLock", {
+    autoRenewLockDurationInMs: sessionIdleTimeoutMs
+  });
+
+  try {
+    // TODO: I just picked a call that didn't grab a message here, just to ensure a session could
+    // actually be obtained. This goes away when we have `await createSessionReceiver()`.
+    await sessionReceiver.getState();
+  } catch (err) {
+    // TODO: when this happens do I still need to close the session receiver? (again, temporary concern)
+    if (_noSessionsAvailable(err)) {
+      console.log(`INFO: no available sessions, sleeping for ${delayWhenNoSessionsAvailableMs}`);
+      await delay(delayWhenNoSessionsAvailableMs);
+      return;
+    }
+
+    throw err;
+  }
+
+  await handlers.processInitialize(sessionReceiver.sessionId);
+
+  const idleTimer = _createIdleTimer(sessionIdleTimeoutMs);
+  let threwError = false;
+
+  sessionReceiver.subscribe(
+    {
+      async processMessage(msg) {
+        idleTimer.refresh();
+        await handlers.processMessage(msg);
+      },
+      async processError(err) {
+        threwError = true;
+        await handlers.processError(err, sessionReceiver.sessionId);
+      }
+    },
+    {
+      abortSignal
+    }
+  );
+
+  await idleTimer.expirationPromise;
+
+  try {
+    await sessionReceiver.close();
+  } catch (err) {
+    await handlers.processError(err, sessionReceiver.sessionId);
+  }
+
+  await handlers.processClose(threwError ? "error" : "idle_timeout", sessionReceiver.sessionId);
+}
+
+async function _roundRobinThroughAvailableSessions(
+  handlers: SessionMessageHandlers,
+  abortSignal: AbortSignalLike
+): Promise<void> {
+  const serviceBusClient = new ServiceBusClient(serviceBusConnectionString);
+  const allPromises = [];
+
+  for (let i = 0; i < maxSessionsToProcessSimultaneously; ++i) {
+    allPromises.push(
+      new Promise(async (_res, rej) => {
+        try {
+          while (!abortSignal.aborted) {
+            await _processNextSession(serviceBusClient, handlers, abortSignal);
+          }
+        } catch (err) {
+          rej(err);
+        }
+      })
+    );
+  }
+
+  // if we get a fatal error in any of the promises we can just shut it all down
+  console.log(`All session promises have been started`);
+  await Promise.race(allPromises);
```

Comment:
What is the reason to wait for the promise returned by `_roundRobinThroughAvailableSessions` to resolve?
Looks like if there was no fatal error from individual sessions, then this promise would never resolve.